library/bignum_new.c

```diff
@@ -0,0 +1,511 @@
+/**
+ *  Internal bignum functions
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_BIGNUM_C)
+
+#include <string.h>
+
+#include "mbedtls/platform_util.h"
+#include "mbedtls/error.h"
+#include "mbedtls/bignum.h"
+#include "bignum_core.h"
+#include "bignum_mod.h"
+#include "bignum_mod_raw.h"
+#include "constant_time_internal.h"
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#define mbedtls_printf      printf
+#define mbedtls_calloc      calloc
+#define mbedtls_free        free
+#endif
+
+#define MPI_VALIDATE_RET( cond )                                       \
+    MBEDTLS_INTERNAL_VALIDATE_RET( cond, MBEDTLS_ERR_MPI_BAD_INPUT_DATA )
+#define MPI_VALIDATE( cond )                                           \
+    MBEDTLS_INTERNAL_VALIDATE( cond )
+
+#define ciL    (sizeof(mbedtls_mpi_uint))   /* chars in limb  */
+#define biL    (ciL << 3)                   /* bits  in limb  */
+#define biH    (ciL << 2)                   /* half limb size */
+
+/*
+ * Convert between bits/chars and number of limbs
+ * Divide first in order to avoid potential overflows
+ */
+#define BITS_TO_LIMBS(i)  ( (i) / biL + ( (i) % biL != 0 ) )
+#define CHARS_TO_LIMBS(i) ( (i) / ciL + ( (i) % ciL != 0 ) )
+
+/*
+ * Count leading zero bits in a given integer
+ */
+size_t mbedtls_mpi_core_clz( const mbedtls_mpi_uint x )
```

Comment:
aren't there often builtins for `clz`?

Reply:
There are, but this is a pre-existing code and we didn't consider rewriting/improving on it.

Thinking about it, I am not sure it is worth using builtins here:

- It would increase code size (as we would need to keep the fallback around)
- Would make the source code more complex
- The impact on overall performance would be negligible

Reply:
We wouldn't need to compile both the builtin version and the plain C version in the same executable, unless the presence of the clz instruction is determined at runtime.

But this is only a minor optimization, and it's a preexisting issue, so now is not the time to worry about it.

Reply:
(came here to post pretty much the same as Gilles: it WOULD increase the size and complexity of the source code, but it WON'T increase code size, and will likely reduce it, and increase performance)

Reply:
See https://en.wikipedia.org/wiki/Find_first_set#Hardware_support - it's on most ARM, sadly excluding Cortex-M0, M0+, M1 and M23

Reply:
It doesn't increase code size if we don't include a fallback, I was just assuming that we would, like we do it with `bswap`.

Do you think it is safe to remove the fallback from there as well?

Reply:
Not sure what you mean by "fallback".

If we can determine at compile time that we have clz/bswap, we use those, and the "fallback" C implementation isn't compiled.

Depending on compiler flags, it's possible to use (e.g.) `__builtin_clz()` and have the compiler just issue a single `clz` instruction, or for the compiler to issue `blx __clzsi2`.

But we can shelve this until later...